File: Search_a_2D_Matrix/Search_a_2D_Matrix.cpp
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        int ROWS = matrix.size();
        int COLS = matrix[0].size();
        int top = 0;
        int bottom = ROWS - 1;

        while (top < bottom)
        {
            int medianRow = top + (bottom - top) / 2;
            if (target > matrix[medianRow][0])
            {
                top = medianRow + 1;
            }
            else
            {
                bottom = medianRow - 1;
            }

            if (matrix[medianRow][0] == target)
            {
                return true;
            }

            if (target > matrix[medianRow][0] && target < matrix[medianRow + 1][0])
            {
                top = medianRow;
                break;
            }
        }
        int left = 0;
        int right = COLS - 1;
        while (left <= right)
        {
            int medianCols = left + (right - left) / 2;
            if (matrix[top][medianCols] == target)
            {
                return true;
            }
            if (target > matrix[top][medianCols])
            {
                left = medianCols + 1;
            }
            else
            {
                right = medianCols - 1;
            }
        }
        return false;
    }
};

File: Search_a_2D_Matrix/Search_a_2D_Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # define coloumns and rows
        ROWS, COLS = len(matrix), len(matrix[0])
        top, bottom = 0, ROWS-1
        # binary search -> bottom & top
        while top < bottom:
            medianRow = (top+bottom)//2
            # greater the last integer of the medianRow
            if target > matrix[medianRow][-1]:
                top = medianRow+1
            elif target < matrix[medianRow][0]:
                bottom = medianRow-1
            else:
                break

        if not (top <= bottom):
            return False

        left, right = 0, COLS-1
        row = (top+bottom)//2
        while left <= right:
            medianCols = (left+right)//2
            if target > matrix[row][medianCols]:
                left = medianCols + 1
            elif target < matrix[row][medianCols]:
                right = medianCols-1
            else:
                return True
        return False
